src/publisher/fragment_logic.py

```diff
@@ -181,24 +187,39 @@ def merge_if_valid(av, quiet=True):
     if invalid and quiet=False, a ValidationError will be raised"""
     return valid(merge_and_preprocess(av), quiet=quiet)
 
-def set_article_json(av, quiet):
+def hash_ajson(merge_result):
+    string = utils.json_dumps(merge_result, indent=None)
+    return hashlib.md5(string.encode('utf-8')).hexdigest()
+
+class Identical(RuntimeError):
+    def __init__(self, msg, av, hashval):
+        super(Identical, self).__init__(msg)
+        self.av = av
+        self.hashval = hashval
+
+# TODO: rename `quiet` to `valid_check` or similar.
+def set_article_json(av, quiet, hash_check=True):
     """updates the article with the result of the merge operation.
     if the result of the merge was valid, the merged result will be saved.
     if invalid, the ArticleVersion instance's article-json will be unset.
     if invalid and quiet=False, a ValidationError will be raised"""
-    log_context = {'article-version': av, 'quiet': quiet}
+    log_context = {'article-version': av, 'quiet': quiet, 'hash_check': hash_check}
     result = merge_if_valid(av, quiet=quiet)
+    newhash, oldhash = hash_ajson(result), av.article_json_hash
+    if hash_check and oldhash == newhash:
+        raise Identical("article data is identical to the article data already stored", av, newhash)
```

Comment:
An exception for a supported flow (idempotence of ingest) is suspicious, can it be substituted with an early `return`?

Reply:
I can raise the exception here or I can return a different value and raise the exception there, but I need to raise something to get the transaction to rollback. Using exceptions as flow control isn't a large sin and this is a delicate area of the codebase - I'd prefer to follow existing patterns and raise the exception.